cpu/nrf52/radio/nrf802154/nrf802154_radio.c

```diff
@@ -0,0 +1,732 @@
+        NRF_RADIO->TXPOWER = RADIO_TXPOWER_TXPOWER_Pos8dBm;
+    }
+    else if (txpower > 1) {
+        NRF_RADIO->TXPOWER = (uint32_t)txpower;
+    }
+    else if (txpower > -1) {
+        NRF_RADIO->TXPOWER = RADIO_TXPOWER_TXPOWER_0dBm;
+    }
+    else if (txpower > -5) {
+        NRF_RADIO->TXPOWER = RADIO_TXPOWER_TXPOWER_Neg4dBm;
+    }
+    else if (txpower > -9) {
+        NRF_RADIO->TXPOWER = RADIO_TXPOWER_TXPOWER_Neg8dBm;
+    }
+    else if (txpower > -13) {
+        NRF_RADIO->TXPOWER = RADIO_TXPOWER_TXPOWER_Neg12dBm;
+    }
+    else if (txpower > -17) {
+        NRF_RADIO->TXPOWER = RADIO_TXPOWER_TXPOWER_Neg16dBm;
+    }
+    else if (txpower > -21) {
+        NRF_RADIO->TXPOWER = RADIO_TXPOWER_TXPOWER_Neg20dBm;
+    }
+    else {
+        NRF_RADIO->TXPOWER = RADIO_TXPOWER_TXPOWER_Neg40dBm;
+    }
+}
+
+static void _disable(void)
+{
+    /* set device into DISABLED state */
+    if (NRF_RADIO->STATE != RADIO_STATE_STATE_Disabled) {
+        NRF_RADIO->EVENTS_DISABLED = 0;
+        NRF_RADIO->TASKS_DISABLE = 1;
+        DEBUG("[nrf802154] Device state: DISABLED\n");
+    }
+}
+
+static void _set_ifs_timer(bool lifs)
+{
+    uint8_t timeout;
+    cfg.ifs = true;
+    if (lifs) {
+        timeout = IEEE802154_LIFS_SYMS;
+    }
+    else {
+        timeout = IEEE802154_SIFS_SYMS;
+    }
+
+    timer_set(NRF802154_TIMER, MAC_TIMER_CHAN_IFS, timeout);
+    timer_start(NRF802154_TIMER);
+}
+
+static int _confirm_set_trx_state(ieee802154_dev_t *dev)
+{
+    (void) dev;
+    int radio_state = NRF_RADIO->STATE;
+    if (radio_state == RADIO_STATE_STATE_TxRu || radio_state == RADIO_STATE_STATE_RxRu ||
+        radio_state == RADIO_STATE_STATE_TxDisable || radio_state == RADIO_STATE_STATE_RxDisable) {
+        return -EAGAIN;
+    }
+    return 0;
+}
+
+static int _request_set_trx_state(ieee802154_dev_t *dev, ieee802154_trx_state_t state)
+{
+    (void) dev;
+
+    if (_state != STATE_IDLE && _state != STATE_RX) {
+        return -EBUSY;
+    }
+
+    _disable();
+
+    /* This will take in worst case 21 us */
+    while (NRF_RADIO->STATE != RADIO_STATE_STATE_Disabled) {};
+
+    switch (state) {
+    case IEEE802154_TRX_STATE_TRX_OFF:
+        _state = STATE_IDLE;
+        break;
+    case IEEE802154_TRX_STATE_RX_ON:
+        NRF_RADIO->PACKETPTR = (uint32_t)rxbuf;
+        NRF_RADIO->TASKS_RXEN = 1;
+        _state = STATE_RX;
+        break;
+    case IEEE802154_TRX_STATE_TX_ON:
+        NRF_RADIO->PACKETPTR = (uint32_t)txbuf;
+        _state = STATE_IDLE;
+        break;
+    }
+
+    return 0;
+}
+
+static void _timer_cb(void *arg, int chan)
+{
+    (void)arg;
+    ieee802154_dev_t *dev = &nrf802154_hal_dev;
+
+    if (chan == MAC_TIMER_CHAN_ACK) {
+        /* Copy sqn */
+        ack[3] = rxbuf[3];
+
+        NRF_RADIO->PACKETPTR = (uint32_t) &ack;
+        NRF_RADIO->TASKS_TXEN = 1;
+        dev->cb(dev, IEEE802154_RADIO_INDICATION_RX_DONE);
+    }
+    else if (chan == MAC_TIMER_CHAN_IFS) {
+        cfg.ifs = false;
```

Comment:
Should this generate a notification to the submac that the next frame can now be transmitted?

Reply:
Do I get it right that the IFS period starts after the TX done event?
Or do you say other radios will only report TX done after the IFS period is over?

(I'm almost certain `at86rf215` won't do any internal calculations w.r.t. timings as I already had do the calculations for all the other timings manually. They'll probably differ between modulations too.)

Reply:
![image](https://user-images.githubusercontent.com/1301112/92747172-3a71f180-f384-11ea-9e71-da038ce1ab25.png)

No delay between the end of RF signal and TXFE on `at86rf215`

Reply:
> Do I get it right that the IFS period starts after the TX done event?
Or do you say other radios will only report TX done after the IFS period is over?

I'm not sure when this event is generated, but I'm sure most radios already handle the IFS (maybe @PeterKietzmann can confirm this). The actually, the `nrf802154` supports hardware IFS but it's disabled in order to use the fast mode (this means, switching between TRX_OFF to TX_ON in 40 us). I think the trade off is totally worth it for this radio.

> They'll probably differ between modulations too.

Yes, it depends on the modulation (it's expressed in symbol times).